Add battle logic with StarWars theme for Fleet and Ship services

gnuheike opened the pull request:
This commit introduces several new services and Interface to handle domain logic for Fleet and Ship entities. This includes new services for handling FleetCombat, ShipDamageProcessing, ShipTargetSelection, Battle, and more. They are essential for defining the actions the entities can take, like causing and receiving damage, selecting targets, and engaging in battles. Additionally, this commit includes a series of associated unit tests for the implemented functionality.

The structured domain logic enables scalability and ongoing support for further feature development. The new additions adhere to appropriate design principles and focus on decoupling dependencies which improves readability and extendability of the code base.

gnuheike committed:
Add battle logic with StarWars theme for Fleet and Ship services

This commit introduces several new services and Interface to handle domain logic for Fleet and Ship entities. This includes new services for handling FleetCombat, ShipDamageProcessing, ShipTargetSelection, Battle, and more. They are essential for defining the actions the entities can take, like causing and receiving damage, selecting targets, and engaging in battles. Additionally, this commit includes a series of associated unit tests for the implemented functionality.

The structured domain logic enables scalability and ongoing support for further feature development. The new additions adhere to appropriate design principles and focus on decoupling dependencies which improves readability and extendability of the code base.

gnuheike committed:
Changed CI to test on PR

Bukharovsi commented on src/Domain/Battle/FleetInterface.php:
```diff
@@ -0,0 +1,10 @@
+<?php
```
Интерфейсы должны лежать там где они используются, а не там где имплементируются. Тут конечно правда, от слов не отказываюсь.

Другое дело что интрефейсы нужны тогда, когда у тебя есть несколько имплементаций (если одна из них тестовая -- то это тоже считается). Но если даже в тестах не заменяем имплеентацию, то мне кажется тогда и интерфейс излишен. 

Я вот для себя какое правило вывел 
 - я не использую слово interface, ну оно как-то врывается в доменый язык и явно там лишнее. Гораздо лаконичнее просто battle, starship и пр
 -  Соотвественно нормальное абстрактное название у меня зарезервинровано под интрефейс. Тот же Fleet к примеру. А теперь переходим к имплементации. И я думаю вот что такого особенного в этой имлементации, что можно к названию Fleet добавить. Если ничего лучше суффикса Impl не придумывается, то я удаляю интерфейс. 

Только с useCase дало сбой. Но об этом отдельно.

+ сам видишь что ты отлично сгруппировал все классы по доменам Battle, Fleet и пр, а FleetInterface почему то в Battle. В общем, мне кажется сократив число таких интерфейсов приложуха только выиграет. 

Вот кстати Еще хороший пример BattleResultRepository -- это интерфейс конечно, верно. Дальше идем к реплизации. Чем она особенная? а Она у нас InMemory, значит будет InMemoryBattleResultRepository - значит пусть интерфейс живет

gnuheike replied:
Спасибо, действительно убрав этот интерфейс "на будущее" стало сильно проще.

Bukharovsi commented on src/Application/UseCase/StoreBattleResult/StoreBattleResult.php:
```diff
@@ -0,0 +1,19 @@
+
+use StarWars\Application\UseCase\ProcessBattle\BattleResult;
+
+class StoreBattleResult
```
Мне кажется интерфейсы из этой директории используются только на уровне UseCase, так что можно их в UseCase и вынести

gnuheike replied:
Не до конца понял, этот UseCase использует интерфейс репозитория для возможности сохранения данных. И сам интерфейс лежит в папке с UseCase, а реализация (InMemory) лежит уже в инфраструктуре. Не мог бы ты пояснить? Спасибо!

Bukharovsi commented on src/Domain/ShipDamageControl/Service/ShipDamageProcessorService.php:
```diff
@@ -0,0 +1,41 @@
+use StarWars\Domain\Ship\Ship;
+use StarWars\Domain\Ship\ShipWeapon;
+
+final class ShipDamageProcessorService implements ShipDamageProcessorServiceInterface
```
Тут Конечно можно сказать что вкусовщина, но в то же время Мартин Фаулер называл это Анемичной Моделью и даже говорил что антипаттерном показывает. 

Суть вот в чем, что у тебя Sheep и Fleet являются объектами хранящими состояние. А вот вся обработка вынесена наружу.
Я бы сделал наоборот.  DSL типа такого
```
val myFleet: Fleet
val sithFleet: Fleet

for (battleShip in myFleet) {
     battleShip.fireRandomly(sithFleet) // и дальше наш корабль пусть выбирает какой корабль он будет отаковать, а дальше корабль ситхов сам пусть считает повреждения. 
Опять таки можно и использовать (и даже в большинстве случаем нужно) вспомогательные классы, ну там стратегия как выбрать цель. Ну скажем Рандомно или же будем палить по той, у которой здоровья меньше. Но все эти товарищи имхо должны инжектится внутрь корабля, Fleet и пр
}
```

gnuheike replied:
Я согласен, не до конца разобрался с моделями и логикой. Перенес эту логику в модель корабля. Звучит логично, что корабль умеет стрелять и сам выбирает цель.

Bukharovsi commented on src/Domain/ShipDamageControl/Service/ShipDamageProcessorService.php:
```diff
@@ -0,0 +1,41 @@
+use StarWars\Domain\Ship\Ship;
+use StarWars\Domain\Ship\ShipWeapon;
+
+final class ShipDamageProcessorService implements ShipDamageProcessorServiceInterface
```
Не могу найти статью, но вообще лучше избегать названий с суффиксами Service, Processor и прочее. Они видть ясности не доабвляют а только дают лишнюю уверенность (ложную). А вот если убрать ссуфиксы и как-то не получается найти нормальнео имя для класса.то в пору и задуматься над тем а стоит ли ему вообще жить

gnuheike replied:
Согласен, почитал про это.

Bukharovsi commented on src/Application/UseCase/ProcessGame/ProcessGame.php:
```diff
@@ -0,0 +1,37 @@
+use StarWars\Application\UseCase\GetSithFleet\GetSithFleet;
+use StarWars\Application\UseCase\ProcessBattle\ProcessBattle;
+
+final class ProcessGame
```
Как правило мы стараемся не вызывать из одного useCase другой, и в целом за раз дергать лучше только один useCase. 

Если есть жаление из одного UseCase вызвать другой, то лучше подумать не стоит ли в домен эту логику вынести. 

А то если разрешит из одного useCase другой дергать, то все может скатится к большим спагеттиобразным useCase и анемичным объектам предметной области

gnuheike replied:
Понял. Убрал это и упростил остальные, чтобы было их логично использовать. Спасибо!

gnuheike committed:
Remove FleetInterface

gnuheike committed:
Remove Unused Interfaces

gnuheike committed:
Removed unused classes

gnuheike committed:
DamageProcessing moved to ship

gnuheike committed:
Remove FleetFactoryInterface and Fleet concept

Replaced the use of Fleet concept with an array, reducing unnecessary complexity. Removed FleetFactoryInterface and related classes that were dependent on it, such as: FleetFactory, GetPlayerFleet, GetSithFleet, ProcessGame, etc. Also, extracted a ShipProviderInterface from ShipsProviderInterface to better highlight its responsibilities. Some test cases related to deleted classes also removed. On the other hand, added some new factory classes to create Ship objects: ShipNameFactory, ShipWeaponFactory, ShipWeaponDamageFactory and FleetFactory.

gnuheike committed:
CS
